src/hotspot/share/opto/mempointer.hpp

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_OPTO_MEMPOINTER_HPP
+#define SHARE_OPTO_MEMPOINTER_HPP
+
+#include "opto/memnode.hpp"
+#include "opto/noOverflowInt.hpp"
+
+// The MemPointer is a shared facility to parse pointers and check the aliasing of pointers,
+// e.g. checking if two stores are adjacent.
+//
+// -----------------------------------------------------------------------------------------
+//
+// MemPointerDecomposedForm:
+//   When the pointer is parsed, it is decomposed into a constant and a sum of summands:
+//
+//     pointer = con + sum(summands)
+//
+//   Where each summand_i in summands has the form:
+//
+//     summand_i = scale_i * variable_i
```

Comment:
I see you treat `scale` as compile time integer value (NoOverflowInt) and not Node. Why is the check/filter for that?

Reply:
I mean: _Where_ is the check/filter for that?

Reply:
The meat of the parsing happens in `MemPointerDecomposedFormParser::parse_sub_expression`.

One good example is this:
![image](https://github.com/user-attachments/assets/a83fd973-6407-45c6-8597-ef51edd7325d)

I added the `default` case now with a comment.

Reply:
okay